### PR TITLE
dcos-launch based aws backend

### PR DESCRIPTION
**Pull request by timaa2k**
This PR adds a dcos-launch based AWS backend to dcos-e2e.
The backend can be plugged in when creating a cluster instead of the Docker backend.
dcos-launch is vendorized to sanitize dependency management.

### CODE REVIEW

File: src/dcos_e2e/cluster.py [hunk 1]
```diff
@@ -117,13 +117,20 @@ def _wait_for_dcos_diagnostics(self) -> None:
         """
         for node in self.masters:
             node.run(
-                args=['/opt/mesosphere/bin/dcos-diagnostics', '--diag'],
+                args=[
```
Comment (adamtheturtle) [inline src/dcos_e2e/cluster.py]: Why these changes? To support 1.9? If so - can this be separate and part of https://jira.mesosphere.com/browse/DCOS_OSS-2299?

File: src/dcos_e2e/cluster.py [hunk 2]
```diff
@@ -22,14 +22,14 @@
...
-    delay=1,
+    delay=5,
 )
 def _wait_for_ssh(node: Node) -> None:
     """
     Retry for up to one minute (arbitrary) until SSH is available on the given
     node.
     """
-    node.run(args=['echo'])
+    node.run(args=['pwd'], log_output_live=True)
```
Comment (timaa2k) [inline src/dcos_e2e/cluster.py]: Note to self: `log_output_live=True` should be removed.

File: src/dcos_e2e/cluster.py [hunk 3]
```diff
@@ -22,14 +22,14 @@
 @retry(
     exceptions=(subprocess.CalledProcessError),
     tries=60,
-    delay=1,
+    delay=5,
```
Comment (adamtheturtle) [inline src/dcos_e2e/cluster.py]: Why?

File: tests/conftest.py [hunk 4]
```diff
@@ -88,6 +89,14 @@ def oss_artifact_url() -> str:
     return 'https://downloads.dcos.io/dcos/stable/dcos_generate_config.sh'
 
 
+@pytest.fixture(scope='session')
+def ee_artifact_url() -> str:
+    """
+    Return the url to a build artifact for DC/OS OSS.
```
Comment (adamtheturtle) [inline tests/conftest.py]: This says "DC/OS OSS" but should be DC/OS Enterprise.

File: src/dcos_e2e/cluster.py (hunk 1 above)
Comment (timaa2k) [reply src/dcos_e2e/cluster.py]: Required for resilience tests. Yes, this is the only change needed to support 1.9 with the AWS backend.

File: docs/source/contributing.rst [hunk 5]
```diff
@@ -157,6 +157,21 @@ To update these links use the following commands, after setting the following en
...
 
+Updating the Amazon Web Services credentials
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Private credentials for Amazon Web Services used by Travis CI.
+
+To update the credentials use the following commands, after setting the following environment variables:
+
+* ``AWS_ACCESS_KEY_ID``
+* ``AWS_SECRET_ACCESS_KEY``
```
Comment (adamtheturtle) [inline docs/source/contributing.rst]: Note where in OneLogin the currently used keys are from.

File: docs/source/aws-backend.rst [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+* ``us-west-2`` US East (Oregon)
+* ``us-east-1`` US East (North Virginia)
+* ``us-east-2`` US East (Ohio)
+* ``eu-cental-1`` EU (Frankfurt)
+* ``eu-west-1`` EU (Ireland)
+
+Restricting access to the cluster
+---------------------------------
+
+The AWS backend takes a parameter ``admin_location``. This parameter restricts the access to the AWS stack from the outside to a particular IP address range.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: One sentence per line.

File: docs/source/aws-backend.rst [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+==============
+
+The AWS backend is used to spin up clusters using EC2 instances on Amazon Web Services, where each instance is a DC/OS node.
+
+.. include:: aws-backend-requirements.rst
+
+DC/OS Installation
+------------------
+
+``Cluster``\ s created by the AWS backend only support installing DC/OS via ``install_dcos_from_url``.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: Let's use ``` :py:class:`~dcos_e2e.cluster.Cluster` ``` here and elsewhere in this document, and similar for other linkable items.

File: docs/source/aws-backend.rst [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+It is important to keep in mind that ``workspace_dir`` is a temporary directory and therefore will be cleaned up after the test.
+If ``workspace_dir`` is unset the AWS backend will create a new temporary directory in a operating system specific location.
+
+Cluster lifetime
+----------------
+
+The cluster lifetime is fixed at a maximum of two hours.
+That is because of a limitation of ``dcos-launch`` which is used under the hood.
+That will most likely change in the future.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: ❓ Link to a JIRA for this?

File: docs/source/aws-backend.rst [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+If the cluster was launched with ``maws`` (Mesosphere temporary AWS credentials) the cluster can be controlled via CCM at ``ccm.mesosphere.com``.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: Use a link to CCM.

File: docs/source/aws-backend.rst [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+This allows for extending the cluster lifetime and also for cleaning up the cluster if anything goes wrong.
+
+EC2 instance types
+------------------
+
+The AWS backend does not offer a choice for EC2 instances.
+Currently it launches ``m4.large`` instances for DC/OS nodes.
+Please direct requests for supporting more instance types towards Adam Dangoor.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: Haha and what am I meant to do with such requests 👅

File: docs/source/aws-backend.rst [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+The AWS backend does currently not support the ``Cluster`` feature of copying files to the DC/OS installer.
+
+Troubleshooting
+---------------
+
+In case of an error during the DC/OS installation the journal from each node will be dumped and downloaded to the folder that the tests were executed in.
+The logs are prefixed with the installation phase that failed, ``preflight``, ``deploy`` or ``postflight``.
+
+When using temporary credentials it is required to pay attention that the credentials are still valid or renewed when deleting a cluster.
+If the credentials are not valid anymore the AWS backend fails to delete the private key pair generated at cluster creation.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: > fails to delete

I guess "does not delete"? Why not?

File: docs/source/aws-backend.rst [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+It is therefore recommended to a periodically renew temporary AWS credentials when executing tests using the AWS backend.
+
+Reference
+---------
+
+.. autoclass:: dcos_e2e.backends._aws.AWS
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: I have updated the Docker backend docs to only document public classes, let's change this to match.

File: docs/source/aws-backend.rst [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+The AWS backend takes a parameter ``admin_location``. This parameter restricts the access to the AWS stack from the outside to a particular IP address range.
+The default value ``'0.0.0.0/0'`` will allow accessing the cluster from anywhere.
+It is recommended to restrict the address range to a subnet including the public IP of the machine executing tests with the AWS backend.
+For example ``<external-ip>/24``.
+
+Accessing cluster nodes
+-------------------------
+
+SSH can be used to access cluster nodes for the purpose of debugging.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: ... if you have set a ``workspace_dir`` on the backend.

File: docs/source/aws-backend.rst [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+The AWS backend generates a key in the ``workspace_dir`` directory under ``ssh/id_rsa``. Adding this key to the ``ssh-agent`` or changing its file permissions to ``400`` will allow for connecting to the cluster via the ``ssh`` command.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: Is that right that it is directly in the ``workspace_dir/ssh`` - isn't it in some random subdirectory of that?

File: tests/test_dcos_e2e/backends/aws/test_aws.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+                extra_config=config,
+                log_output_live=True,
+            )
+
+            cluster.wait_for_dcos_ee(
+                superuser_username=superuser_username,
+                superuser_password=superuser_password,
+            )
+
+            def private_hosts(nodes: Set[Node]) -> str:
```
Comment (adamtheturtle) [inline tests/test_dcos_e2e/backends/aws/test_aws.py]: I'd expect to not have to set `MASTER_HOSTS`, `SLAVE_HOSTS` or `PUBLIC_SLAVE_HOSTS`. Do  you think that `cluster.run_integration_tests` should be changed to use the private IP address by default?

File: tests/test_dcos_e2e/backends/aws/test_aws.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+            cluster.run_integration_tests(
+                pytest_command=['pytest', '-vvv', '-s', '-x', 'test_tls.py'],
+                env={
+                    'DCOS_LOGIN_UNAME': superuser_username,
+                    'DCOS_LOGIN_PW': superuser_password,
+                    'DCOS_SSL_ENABLED': 'true',
+                    'MASTER_HOSTS': private_hosts(cluster.masters),
+                    'SLAVE_HOSTS': private_hosts(cluster.agents),
+                    'PUBLIC_SLAVE_HOSTS': private_hosts(cluster.public_agents),
+                    'DCOS_DNS_ADDRESS': 'https://' + str(master_ip),
```
Comment (adamtheturtle) [inline tests/test_dcos_e2e/backends/aws/test_aws.py]: Should `DCOS_DNS_ADDRESS` be set by `cluster.run_integration_tests`? How about `DCOS_SSL_ENABLED`? I think that we should at least document these requirements somewhere.

File: docs/source/aws-backend-requirements.rst [hunk 17]
```diff
@@ -0,0 +1,39 @@
+Requirements
+------------
+
+Amazon Web Services
+~~~~~~~~~~~~~~~~~~~
+
+An Amazon Web Services account with sufficient funds must be available.
+
+The AWS credentials for the account must be present either in the environment as environment variables or in the default file system location under ``~/.aws/credentials``.
```
Comment (adamtheturtle) [inline docs/source/aws-backend-requirements.rst]: For ``~/.aws/credentials``, use the [`:file:` directive](http://www.sphinx-doc.org/en/stable/markup/inline.html#role-file).

File: src/dcos_e2e/_vendor/dcos_launch/onprem.py [hunk 18]
```diff
@@ -109,7 +109,7 @@ def get_completed_onprem_config(self) -> typing.Tuple[dict, str]:
             # use a sensible default
             shutil.copyfile(
                 pkg_resources.resource_filename(
-                    'dcos_launch', script_hyphen + '/{}.sh'.format(self.config['platform'])),
+                    'dcos_e2e._vendor.dcos_launch', script_hyphen + '/{}.sh'.format(self.config['platform'])),
```
Comment (adamtheturtle) [inline src/dcos_e2e/_vendor/dcos_launch/onprem.py]: We should document this manual override in `contributing.rst` and link to a JIRA for upstreaming it.

File: tests/test_dcos_e2e/backends/aws/test_aws.py (hunk 15 above)
Comment (timaa2k) [reply tests/test_dcos_e2e/backends/aws/test_aws.py]: Yes, indeed.

File: tests/test_dcos_e2e/backends/aws/test_aws.py (hunk 16 above)
Comment (timaa2k) [reply tests/test_dcos_e2e/backends/aws/test_aws.py]: This is linked to using the `private_ip_address`.

File: docs/source/contributing.rst (hunk 5 above)
Comment (timaa2k) [reply docs/source/contributing.rst]: Noted below.

File: src/dcos_e2e/cluster.py (hunk 3 above)
Comment (timaa2k) [reply src/dcos_e2e/cluster.py]: Arbitrary, it works. Might also work with 1 sec though.

Comment (adamtheturtle) [reply src/dcos_e2e/cluster.py]: If this is a real fix for the issue let's take it out to a new PR / JIRA.

File: docs/source/aws-backend.rst (hunk 11 above)
Comment (timaa2k) [reply docs/source/aws-backend.rst]: Documented in the `_aws/__init__.py`.

File: tests/test_dcos_e2e/backends/aws/test_aws.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+            master_ip = next(iter(cluster.masters)).private_ip_address
+
+            # No error is raised with a successful command.
+            cluster.run_integration_tests(
+                pytest_command=['pytest', '-vvv', '-s', '-x', 'test_tls.py'],
+                env={
+                    'DCOS_LOGIN_UNAME': superuser_username,
+                    'DCOS_LOGIN_PW': superuser_password,
+                    'DCOS_SSL_ENABLED': 'true',
```
Comment (adamtheturtle) [inline tests/test_dcos_e2e/backends/aws/test_aws.py]: Still to do:

Document these options or move them to `cluster.run_integration_tests`.

File: src/dcos_e2e/backends/_aws/__init__.py [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+        self.launcher.config['dcos_config'] = {**dcos_config, **extra_config}
+
+        # The ``wait`` method starts the actual DC/OS installation process.
+        # We do not use ``self.launcher.wait()`` here because at the time of
+        # writing it does both, waiting for the AWS stack and installing
+        # DC/OS. This is desired to be changed by the dcos-launch team.
+        # https://jira.mesosphere.com/browse/DCOS-21660
+        try:
+            AbstractOnpremLauncher.wait(self.launcher)  # type: ignore
+        except (KeyboardInterrupt, Exception):
```
Comment (adamtheturtle) [inline src/dcos_e2e/backends/_aws/__init__.py]: Is it possible that someone will want to interrupt the install of DC/OS and not destroy the cluster?

What do you think of moving this to `cluster.py`? In a separate PR?

File: src/dcos_e2e/cluster.py [hunk 21]
```diff
@@ -22,14 +22,14 @@
...
 )
 def _wait_for_ssh(node: Node) -> None:
     """
     Retry for up to one minute (arbitrary) until SSH is available on the given
     node.
     """
-    node.run(args=['echo'])
+    node.run(args=['pwd'])
```
Comment (adamtheturtle) [inline src/dcos_e2e/cluster.py]: Similarly, I think that this does nothing.

File: src/dcos_e2e/backends/_aws/__init__.py (hunk 20 above)
Comment (timaa2k) [reply src/dcos_e2e/backends/_aws/__init__.py]: This could indeed be moved to cluster.py.
A half-provisioned cluster remains in an undefined state and from my point of view is useless to keep around.

File: docs/source/aws-backend.rst [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+The default value ``'0.0.0.0/0'`` will allow accessing the cluster from anywhere.
+It is recommended to restrict the address range to a subnet including the public IP of the machine executing tests with the AWS backend.
+For example ``<external-ip>/24``.
+
+Accessing cluster nodes
+-----------------------
+
+SSH can be used to access cluster nodes for the purpose of debugging if :paramref:`~dcos_e2e.backends.AWS.workspace_dir` is set.
+The AWS backend generates a key in the :paramref:`~dcos_e2e.backends.AWS.workspace_dir` directory under ``ssh/id_rsa``.
+Adding this key to the ``ssh-agent`` or changing its file permissions to ``400`` will allow for connecting to the cluster via the ``ssh`` command.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: > or changing its file permissions to ``400``

Surely there's some other step apart from just changing the file permissions.

File: docs/source/aws-backend.rst [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+The SSH user depends on the :paramref:`~dcos_e2e.backends.AWS.linux_distribution` given to the :py:class:`~dcos_e2e.backends.AWS` backend.
+For :py:obj:`~dcos_e2e.distributions.Distribution.CENTOS_7` that is ``centos``.
+
+It is important to keep in mind files in the given :paramref:`~dcos_e2e.backends.AWS.workspace_dir` are temporary and are removed up when the cluster is destroyed.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: > removed up when the

I guess "cleaned up when" or "removed when"

File: docs/source/aws-backend.rst [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+If :paramref:`~dcos_e2e.backends.AWS.workspace_dir` is unset the :py:class:`~dcos_e2e.backends.AWS` backend will create a new temporary directory in an operating system specific location.
+
+Cluster lifetime
+----------------
+
+The cluster lifetime is fixed at a maximum of two hours.
+That is because of a limitation of `dcos-launch`_ which is used under the hood.
+That will most likely change in the future.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: Is there a link to a JIRA issue for this? If so, link to it. Else, why do you think this?

File: docs/source/aws-backend.rst [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+
+If the cluster was launched with ``maws`` (Mesosphere temporary AWS credentials) the cluster can be controlled via `CCM`_.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: Change mentions of ``maws`` to links to maws.

File: docs/source/aws-backend.rst [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+This allows for extending the cluster lifetime and also for cleaning up the cluster if anything goes wrong.
+
+EC2 instance types
+------------------
+
+The AWS backend does not offer a choice for EC2 instances.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: > a choice for EC2 instances

`s/for/of`.

However, I'd just remove this sentence.

File: docs/source/aws-backend.rst [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+Currently it launches ``m4.large`` instances for DC/OS nodes.
+
+Unsupported features
+--------------------
+
+The AWS backend does currently not support the :py:class:`~dcos_e2e.cluster.Cluster` feature of copying files to the DC/OS installer.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: Use a `:paramref:` to link to the exact unsupported parameter.

File: docs/source/aws-backend.rst [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+
+Troubleshooting
+---------------
+
+In case of an error during the DC/OS installation the journal from each node will be dumped and downloaded to the folder that the tests were executed in.
+The logs are prefixed with the installation phase that failed, ``preflight``, ``deploy`` or ``postflight``.
+
+When using temporary credentials it is required to pay attention that the credentials are still valid or renewed when destroying a cluster.
+If the credentials are not valid anymore the AWS backend does not delete the public/private key pair generated during cluster creation.
+It is therefore recommended to a periodically renew temporary AWS credentials when executing tests using the AWS backend.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: > It is therefore recommended to a periodically

Unnecessary "a ".

File: docs/source/aws-backend.rst [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+
+In rare cases it might also happen that a AWS stack deployment fails with the message ``ROLLBACK_IN_PROGRESS``.
+In that case one of the EC2 instances failed to come up. Re-running the test is the only option then.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: New sentence new line.

"Re-running the test" - let's not talk about tests here - this is a library for creating and managing an AWS DC/OS cluster. It might be used in non-test environments.

File: src/dcos_e2e/backends/_aws/__init__.py [hunk 30]
```diff
@@ -0,0 +1,342 @@
...
+    ) -> None:
+        """
+        Create an AWS cluster.
+
+        Args:
+            masters: The number of master nodes to create.
+            agents: The number of agent nodes to create.
+            public_agents: The number of public agent nodes to create.
+            files_to_copy_to_installer: Pairs of host paths to paths on the
+                installer node. This must be empty for as it is not currently
```
Comment (adamtheturtle) [inline src/dcos_e2e/backends/_aws/__init__.py]: > This must be empty for as

Remove "for".

File: src/dcos_e2e/backends/_aws/__init__.py [hunk 31]
```diff
@@ -0,0 +1,342 @@
...
+            'exhibitor_storage_backend': 'static',
+        }
+
+        # Validate the preliminary dcos-launch config.
+        validated_launch_config = config.get_validated_config(
+            user_config=launch_config,
+            config_dir=str(self._path),
+        )
+
+        # Also DcosCloudformationLauncher
```
Comment (adamtheturtle) [inline src/dcos_e2e/backends/_aws/__init__.py]: > Also DcosCloudformationLauncher
> Get a OnpremLauncher object

This is not a clear comment.

File: src/dcos_e2e/backends/_aws/__init__.py [hunk 32]
```diff
@@ -0,0 +1,342 @@
...
+        # First we create a preliminary dcos-config inside the
+        # dcos-launch config to pass the config validation step.
+        launch_config['dcos_config'] = {
+            'cluster_name': unique,
+            'resolvers': ['10.10.0.2', '8.8.8.8'],
+            'master_discovery': 'static',
+            'exhibitor_storage_backend': 'static',
+        }
+
+        # Validate the preliminary dcos-launch config.
```
Comment (adamtheturtle) [inline src/dcos_e2e/backends/_aws/__init__.py]: More importantly it adds to the config, right?

File: docs/source/contributing.rst [hunk 33]
```diff
@@ -157,6 +158,23 @@ To update these links use the following commands, after setting the following en
     travis env set --repo mesosphere/dcos-e2e EE_1_10_ARTIFACT_URL $EE_1_10_ARTIFACT_URL
     travis env set --repo mesosphere/dcos-e2e EE_1_11_ARTIFACT_URL $EE_1_11_ARTIFACT_URL
 
+Updating the Amazon Web Services credentials
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Private credentials for Amazon Web Services used by Travis CI.
```
Comment (adamtheturtle) [inline docs/source/contributing.rst]: > Amazon Web Services used by Travis CI.

"are used".

**Comment by adamtheturtle** [comment]
I love it ❤️ 

What I need before merging:
* Address minor comments
* Wait for the new tests to pass, again
* Assess the code coverage. We want 100% but have a known error (I'm working on that) so just dig into the code coverage and find 100% for all the changed files.

Then optionally ask for another review but if not, let's merge!

**Comment by adamtheturtle** [comment]
Let's also resolve https://github.com/mesosphere/dcos-e2e/pull/645/files#r177473599 - it might add a 4 second delay to a very quick process (starting nodes on Docker).

File: docs/source/aws-backend.rst [hunk 34]
```diff
@@ -0,0 +1,94 @@
...
+
+In case of an error during the DC/OS installation the journal from each node will be dumped and downloaded to the folder that the tests were executed in.
+The logs are prefixed with the installation phase that failed, ``preflight``, ``deploy`` or ``postflight``.
+
+When using temporary credentials it is required to pay attention that the credentials are still valid or renewed when destroying a cluster.
+If the credentials are not valid anymore the AWS backend does not delete the public/private key pair generated during cluster creation.
+It is therefore recommended to periodically renew temporary AWS credentials when executing tests using the AWS backend.
+
+In rare cases it might also happen that a AWS stack deployment fails with the message ``ROLLBACK_IN_PROGRESS``.
+In that case one of the EC2 instances failed to come up. Starting a new cluster is the only option then.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: one of > at least one of?

File: docs/source/aws-backend.rst [hunk 35]
```diff
@@ -0,0 +1,94 @@
...
+
+It is important to keep in mind files in the given :paramref:`~dcos_e2e.backends.AWS.workspace_dir` are temporary and are removed when the cluster is destroyed.
+If :paramref:`~dcos_e2e.backends.AWS.workspace_dir` is unset the :py:class:`~dcos_e2e.backends.AWS` backend will create a new temporary directory in an operating system specific location.
+
+Cluster lifetime
+----------------
+
+The cluster lifetime is fixed at two hours.
+
+If the cluster was launched with `maws`_ (Mesosphere temporary AWS credentials) the cluster can be controlled via `CCM`_.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: ❓ Why define `maws` here and not above? Personally I'd just not bother with the definition now that it is a link.

Comment (adamtheturtle) [reply docs/source/aws-backend.rst]: "not above" = not when you first mention `maws` in this document.

File: tests/test_dcos_e2e/backends/aws/test_aws.py [hunk 36]
```diff
@@ -0,0 +1,133 @@
...
+            )
+
+            cluster.wait_for_dcos_ee(
+                superuser_username=superuser_username,
+                superuser_password=superuser_password,
+            )
+
+            # No error is raised with a successful command.
+            cluster.run_integration_tests(
+                pytest_command=['pytest', '-vvv', '-s', '-x', 'test_tls.py'],
```
Comment (adamtheturtle) [inline tests/test_dcos_e2e/backends/aws/test_aws.py]: Add a comment that this happens to be a test module that does not require >= 1 of any node type.

File: tests/test_dcos_e2e/backends/aws/test_aws.py [hunk 37]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        with Cluster(
+            cluster_backend=AWS(linux_distribution=linux_distribution),
+            masters=1,
+        ) as cluster:
+
+            (master, ) = cluster.masters
+
+            result = master.run(
```
Comment (adamtheturtle) [inline tests/test_dcos_e2e/backends/aws/test_aws.py]: Why run `echo test` at all? Surely this is covered by `cluster.run_integration_tests`. If it is but you want to narrow the cause of any issue, add a comment to that affect.

File: src/dcos_e2e/backends/_aws/__init__.py [hunk 38]
```diff
@@ -0,0 +1,342 @@
...
+
+
+class AWS(ClusterBackend):
+    """
+    A record of an AWS backend which can be used to create clusters.
+    """
+
+    def __init__(
+        self,
+        aws_region: str = 'us-west-2',
```
Comment (adamtheturtle) [inline src/dcos_e2e/backends/_aws/__init__.py]: This does not match what we recommend for cost - why?

If the reason for this region is cost, can you add a comment?

File: src/dcos_e2e/backends/_aws/__init__.py [hunk 39]
```diff
@@ -0,0 +1,342 @@
...
+        admin_location: str = '0.0.0.0/0',
```
Comment (adamtheturtle) [inline src/dcos_e2e/backends/_aws/__init__.py]: Can we test for what the default region and admin location are?

File: docs/source/aws-backend.rst [hunk 40]
```diff
@@ -0,0 +1,93 @@
...
+This is because the installation method employs a bootstrap node that directly downloads the :paramref:`~dcos_e2e.cluster.Cluster.install_dcos_from_url.build_artifact` from the specified URL.
+
+:py:class:`~dcos_e2e.node.Node`\ s of :py:class:`~dcos_e2e.cluster.Cluster`\ s created by the :py:class:`~dcos_e2e.backends.AWS` backend distinguish between :py:attr:`~dcos_e2e.node.Node.public_ip_address` and :py:attr:`~dcos_e2e.node.Node.private_ip_address`.
+The :py:attr:`~dcos_e2e.node.Node.private_ip_address` refers to the internal network of the AWS stack which is also used by DC/OS internally.
+The :py:attr:`~dcos_e2e.node.Node.public_ip_address` allows for reaching AWS EC2 instances from the outside e.g. from the ``dcos-e2e`` testing environment.
+
+AWS Regions
+-----------
+
+When launching a cluster with Amazon Web Services there are a number of different regions to choose from where the cluster is launched.
```
Comment (adamtheturtle) [inline docs/source/aws-backend.rst]: ❓ Change `regions` to a `:paramref:`.

File: .travis.yml [hunk 41]
```diff
@@ -13,6 +13,8 @@ python:
 env:
   matrix:
   - TEST_PATTERN=tests/test_cli
+  - TEST_PATTERN=tests/test_dcos_e2e/backends/aws/test_aws.py::TestUnsupported
```
Comment (adamtheturtle) [inline .travis.yml]: Missing `TestDefaults`.

**Comment by adamtheturtle** [comment]
Follow-ups:
* Test that If I set a custom `region`, this has an affect on the region.
* If a region is given which is not a "real" region, error nicely (hopefully `boto` can help us with that)
* If I set a custom `admin_location`, that makes a difference

If you agree, please create a JIRA issue. I think we should merge this ASAP and then iterate.

File: src/dcos_e2e/_vendor/dcos_launch/onprem.py (hunk 18 above)
Comment (adamtheturtle) [reply src/dcos_e2e/_vendor/dcos_launch/onprem.py]: (Tim did this)

File: src/dcos_e2e/backends/_aws/__init__.py (hunk 39 above)
Comment (adamtheturtle) [reply src/dcos_e2e/backends/_aws/__init__.py]: (Tim did this)

File: src/dcos_e2e/backends/_aws/__init__.py (hunk 32 above)
Comment (adamtheturtle) [reply src/dcos_e2e/backends/_aws/__init__.py]: This is addressed below.

File: tests/test_dcos_e2e/backends/aws/test_aws.py (hunk 36 above)
Comment (adamtheturtle) [reply tests/test_dcos_e2e/backends/aws/test_aws.py]: This is done in the docstring now.
